### PR TITLE
feature/address suggestions

### PR DESCRIPTION
**Pull request by MaxAstin**

### CODE REVIEW

File: app/src/main/java/com/bunbeauty/papakarlo/common/ui/element/textfield/FoodDeliveryBaseTextField.kt [hunk 1]
```diff
@@ -52,7 +55,9 @@ fun FoodDeliveryBaseTextField(
             )
         },
         trailingIcon = {
-            if (value.isNotEmpty()) {
+            if (isLoading) {
```
Comment (xidealo) [inline app/src/main/java/com/bunbeauty/papakarlo/common/ui/element/textfield/FoodDeliveryBaseTextField.kt:58]: lets make when with else branch

File: app/src/main/java/com/bunbeauty/papakarlo/common/ui/element/textfield/FoodDeliveryBaseTextField.kt [hunk 2]
```diff
@@ -112,7 +117,9 @@ fun FoodDeliveryBaseTextField(
                 )
             },
             trailingIcon = {
-                if (value.text.isNotEmpty()) {
+                if (isLoading) {
```
Comment (xidealo) [inline app/src/main/java/com/bunbeauty/papakarlo/common/ui/element/textfield/FoodDeliveryBaseTextField.kt:120]: same "when"

File: app/src/main/java/com/bunbeauty/papakarlo/common/ui/element/textfield/FoodDeliveryBaseTextField.kt [hunk 3]
```diff
@@ -148,3 +155,27 @@ private fun FoodDeliveryTextBaseFieldPreview() {
         onValueChange = {}
     )
 }
+
+@ExperimentalComposeUiApi
+@Preview
+@Composable
+private fun FoodDeliveryTextBaseFieldWithLoadingPreview() {
+    FoodDeliveryBaseTextField(
```
Comment (xidealo) [inline app/src/main/java/com/bunbeauty/papakarlo/common/ui/element/textfield/FoodDeliveryBaseTextField.kt:163]: do we need use Theme { }?

File: app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt [hunk 4]
```diff
@@ -24,78 +21,77 @@ import androidx.compose.ui.res.stringResource
...
             actionButton = {
-                if (createAddressState.state is CreateAddressState.State.Success) {
-                    LoadingButton(
-                        modifier = Modifier.padding(horizontal = FoodDeliveryTheme.dimensions.mediumSpace),
-                        textStringId = R.string.action_create_address_save,
-                        isLoading = createAddressState.isCreateLoading
-                    ) {
-                        viewModel.onCreateAddressClicked()
+                LoadingButton(
+                    modifier = Modifier.padding(horizontal = FoodDeliveryTheme.dimensions.mediumSpace),
```
Comment (xidealo) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt:52]: dp

File: app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt [hunk 5]
```diff
@@ -24,78 +21,77 @@ import androidx.compose.ui.res.stringResource
...
                     }
-                }
+                )
             }
         ) {
-            when (val state = createAddressState.state) {
-                is CreateAddressState.State.Success -> CreateAddressSuccessScreen(createAddressState)
-                is CreateAddressState.State.Loading -> LoadingScreen()
-                is CreateAddressState.State.Error -> CreateAddressErrorScreen(state.throwable)
+            CreateAddressSuccessScreen(
```
Comment (xidealo) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt:61]: what about crossfade?

File: app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt [hunk 6]
```diff
@@ -112,115 +108,92 @@ class CreateAddressFragment : BaseFragmentWithSharedViewModel(R.layout.layout_co
...
+                        value = viewState.street,
                         labelStringId = R.string.hint_create_address_street,
-                        onValueChange = { value ->
-                            streetText = value
-                            viewModel.onStreetTextChanged(value)
+                        onValueChange = { street ->
+                            onAction(CreateAddress.Action.StreetTextChange(street = street))
                         },
-                        errorMessageId = if (state.hasStreetError) {
+                        errorMessageId = if (viewState.hasStreetError) {
```
Comment (xidealo) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt:135]: you could create viewState and map to streetErrorMessage : Int?
and same for other fields

File: app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt [hunk 7]
```diff
@@ -112,115 +108,92 @@ class CreateAddressFragment : BaseFragmentWithSharedViewModel(R.layout.layout_co
...
+                        value = viewState.house,
                         labelStringId = R.string.hint_create_address_house,
                         onValueChange = { value ->
-                            houseText = value
-                            viewModel.onHouseTextChanged(value)
+                            onAction(CreateAddress.Action.HouseTextChange(house = value))
                         },
                         maxSymbols = 5,
-                        errorMessageId = if (state.hasHouseError) {
+                        errorMessageId = if (viewState.hasHouseError) {
```
Comment (xidealo) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt:152]: avoid this if

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -109,15 +109,11 @@
     <string name="hint_create_address_floor">Этаж</string>
     <string name="hint_create_address_comment">Комментарий для курьера</string>
     <string name="msg_create_address_created">Адрес добавлен</string>
-    <string name="error_create_address_fail">Не удалось создать адрес</string>
     <string name="error_create_address_street">Выберите название улицы из списка</string>
     <string name="error_create_address_house">Укажите номер дома</string>
-    <string name="action_create_address_save">Сохранить</string>
+    <string name="error_create_address_fail">Не удалось создать адрес</string>
```
Comment (xidealo) [inline app/src/main/res/values/strings.xml:114]: создать -> добавить?

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/data/mapper/user_address/UserAddressMapper.kt [hunk 9]
```diff
@@ -40,9 +38,9 @@ class UserAddressMapper(
     fun toUserAddressEntity(addressServer: AddressServer): UserAddressEntity {
         return UserAddressEntity(
             uuid = addressServer.uuid,
-            streetUuid = addressServer.street.uuid,
-            streetName = addressServer.street.name,
-            cityUuid = addressServer.street.cityUuid,
+            streetUuid = "",
```
Comment (xidealo) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/data/mapper/user_address/UserAddressMapper.kt:41]: is it correct?

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/data/repository/SuggestionRepository.kt [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+import com.bunbeauty.shared.domain.repo.SuggestionRepo
+import com.bunbeauty.shared.extension.dataOrNull
+
+class SuggestionRepository(
+    private val networkConnector: NetworkConnector,
+) : SuggestionRepo {
+
+    private val cache: MutableMap<String, List<Suggestion>> = mutableMapOf()
+
+    override suspend fun getSuggestionList(token: String, query: String, cityUuid: String): List<Suggestion>? {
```
Comment (xidealo) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/data/repository/SuggestionRepository.kt:14]: nullable list? 
can we use empty list for case with null?

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/data/repository/SuggestionRepository.kt [hunk 11]
```diff
@@ -0,0 +1,36 @@
+package com.bunbeauty.shared.data.repository
+
+import com.bunbeauty.shared.data.network.api.NetworkConnector
+import com.bunbeauty.shared.domain.model.Suggestion
+import com.bunbeauty.shared.domain.repo.SuggestionRepo
+import com.bunbeauty.shared.extension.dataOrNull
+
+class SuggestionRepository(
+    private val networkConnector: NetworkConnector,
+) : SuggestionRepo {
```
Comment (xidealo) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/data/repository/SuggestionRepository.kt:10]: why not CacheListRepository

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/domain/feature/address/GetSuggestionsUseCase.kt [hunk 12]
```diff
@@ -0,0 +1,25 @@
+package com.bunbeauty.shared.domain.feature.address
+
+import com.bunbeauty.shared.DataStoreRepo
+import com.bunbeauty.shared.domain.exeptions.DataNotFoundException
+import com.bunbeauty.shared.domain.exeptions.NoSelectedCityUuidException
+import com.bunbeauty.shared.domain.exeptions.NoTokenException
+import com.bunbeauty.shared.domain.model.Suggestion
+import com.bunbeauty.shared.domain.repo.SuggestionRepo
+
+class GetSuggestionsUseCase(
```
Comment (xidealo) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/domain/feature/address/GetSuggestionsUseCase.kt:10]: Need tests

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/create_address/CreateAddressViewModel.kt [hunk 13]
```diff
@@ -1,185 +1,259 @@
...
+    @OptIn(FlowPreview::class)
+    private fun observeStreetChanges() {
+        mutableDataState.map { state ->
+            CreateAddress.StreetField(
+                street = state.street,
+                isFocused = state.streetFocused,
+                selectedSuggestion = state.selectedStreetSuggestion,
+            )
+        }.distinctUntilChanged()
+            .debounce(1_000)
```
Comment (xidealo) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/create_address/CreateAddressViewModel.kt:217]: magic nubmer

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/create_address/CreateAddressViewModel.kt [hunk 14]
```diff
@@ -1,185 +1,259 @@
...
+        suggestionsJob?.cancel()
+        suggestionsJob = sharedScope.launchSafe(
+            block = {
+                setState {
+                    copy(isSuggestionLoading = true)
+                }
+                val suggestionList = getSuggestionsUseCase(query = query)
+                setState {
+                    copy(
+                        streetSuggestionList = suggestionList.map { suggestion ->
```
Comment (xidealo) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/create_address/CreateAddressViewModel.kt:238]: extract function

**Comment by xidealo** [comment]
BaseComposeFragment add remember for viewModel lamda pls

File: app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt (hunk 5 above)
Comment (MaxAstin) [reply app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt:61]: no crossfade

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/data/repository/SuggestionRepository.kt (hunk 10 above)
Comment (MaxAstin) [reply shared/src/commonMain/kotlin/com/bunbeauty/shared/data/repository/SuggestionRepository.kt:14]: empty list is a valid response - everything works fine, but nothing was found
null means something went wrong

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/data/repository/SuggestionRepository.kt (hunk 11 above)
Comment (MaxAstin) [reply shared/src/commonMain/kotlin/com/bunbeauty/shared/data/repository/SuggestionRepository.kt:10]: 1. we don't store Suggestions locally 
2. we are using Map for cache, not List

File: app/src/main/java/com/bunbeauty/papakarlo/common/ui/element/textfield/FoodDeliveryBaseTextField.kt (hunk 1 above)
Comment (MaxAstin) [reply app/src/main/java/com/bunbeauty/papakarlo/common/ui/element/textfield/FoodDeliveryBaseTextField.kt]: ✅ fixed

File: app/src/main/java/com/bunbeauty/papakarlo/common/ui/element/textfield/FoodDeliveryBaseTextField.kt (hunk 2 above)
Comment (MaxAstin) [reply app/src/main/java/com/bunbeauty/papakarlo/common/ui/element/textfield/FoodDeliveryBaseTextField.kt]: ✅ fixed

File: app/src/main/java/com/bunbeauty/papakarlo/common/ui/element/textfield/FoodDeliveryBaseTextField.kt (hunk 3 above)
Comment (MaxAstin) [reply app/src/main/java/com/bunbeauty/papakarlo/common/ui/element/textfield/FoodDeliveryBaseTextField.kt]: ✅ fixed

File: app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt (hunk 4 above)
Comment (MaxAstin) [reply app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt]: ✅ fixed

File: app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt (hunk 6 above)
Comment (MaxAstin) [reply app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt]: ✅ fixed

File: app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt (hunk 7 above)
Comment (MaxAstin) [reply app/src/main/java/com/bunbeauty/papakarlo/feature/address/screen/createaddress/CreateAddressFragment.kt]: ✅ fixed

File: app/src/main/res/values/strings.xml (hunk 8 above)
Comment (MaxAstin) [reply app/src/main/res/values/strings.xml]: ✅ fixed

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/data/mapper/user_address/UserAddressMapper.kt (hunk 9 above)
Comment (MaxAstin) [reply shared/src/commonMain/kotlin/com/bunbeauty/shared/data/mapper/user_address/UserAddressMapper.kt]: It's correct. We don't need this field anymore. 
✅ fixed - remove it from Address table

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/domain/feature/address/GetSuggestionsUseCase.kt (hunk 12 above)
Comment (MaxAstin) [reply shared/src/commonMain/kotlin/com/bunbeauty/shared/domain/feature/address/GetSuggestionsUseCase.kt:10]: ✅ added

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/create_address/CreateAddressViewModel.kt (hunk 13 above)
Comment (MaxAstin) [reply shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/create_address/CreateAddressViewModel.kt]: ✅ fixed

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/create_address/CreateAddressViewModel.kt (hunk 14 above)
Comment (MaxAstin) [reply shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/create_address/CreateAddressViewModel.kt]: ✅ fixed

**Pull request merged**